src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,456 @@
+    uint256             constant  RAY  = 10 ** 27;
+
+    VatLike      public immutable vat;
+    DaiJoinLike  public immutable daiJoin;
+    address      public           vow;
+    EndLike      public           end;
+
+    struct Ilk {
+        D3MPoolLike pool;   // Access external pool and holds balances
+        D3MPlanLike plan;   // How we calculate target debt
+        uint256     tau;    // Time until you can write off the debt [sec]
+        uint256     culled; // Debt write off triggered
+        uint256     tic;    // Timestamp when the pool is caged
+    }
+    mapping (bytes32 => Ilk) public ilks;
+    uint256                  public live = 1;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed ilk, bytes32 indexed what, address data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed what, address data);
+    event Wind(bytes32 indexed ilk, uint256 amount);
+    event Unwind(bytes32 indexed ilk, uint256 amount);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Cage();
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk);
+    event Uncull(bytes32 indexed ilk);
+    event Quit(bytes32 indexed ilk, address indexed usr);
+    event Exit(bytes32 indexed ilk, address indexed usr, uint256 amt);
+
+    constructor(address vat_, address daiJoin_) public {
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        TokenLike(DaiJoinLike(daiJoin_).dai()).approve(daiJoin_, type(uint256).max);
+        VatLike(vat_).hope(daiJoin_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+
+        if (what == "vow") vow = data;
+        else if (what == "end") end = EndLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        if (what == "tau" ) {
+            require(live == 1, "DssDirectDepositHub/hub-not-live");
+            require(ilks[ilk].tic == 0, "DssDirectDepositHub/join-not-live");
```

Comment:
Change error to `DssDirectDepositHub/pool-not-live` for consistency.

Also these requires can be moved out of the if block for consistency with other `file` functions.